inc/cppxaml/XamlWindow.h

```diff
@@ -0,0 +1,256 @@
+        /// <param name="markup">
+        /// XAML markup string to use to create the UI.
+        /// </param>
+        /// <param name="c"></param>
+        /// <returns></returns>
+        static XamlWindow& Make(PCWSTR id, std::wstring_view markup, AppController* c = nullptr) {
+            XamlWindow xw(id);
+            xw.controller = c;
+            xw.m_markup = markup;
+            xw.getUI = [](const XamlWindow& xw) {
+                return winrt::Windows::UI::Xaml::Markup::XamlReader::Load(xw.m_markup)
+                    .as<winrt::Windows::UI::Xaml::UIElement>();
+            };
+            const auto& entry = s_windows.emplace(id, std::move(xw));
+            return entry.first->second;
+        }
+
+        /// <summary>
+        /// Creates a window that hosts XAML UI. The UI will be provided by the app.
+        /// </summary>
+        /// <param name="id"></param>
+        /// <param name="getUI">
+        /// App-provided function to create UI. This can e.g. instantiate XAML types programmatically.
+        /// <example><c>auto& xw = cppxaml::XamlWindow::Make(L"Foo", [](auto&amp;...) { return winrt::Windows::UI::Xaml::Controls::Button(); });</c></example>
+        /// </param>
+        /// <param name="c"></param>
+        /// <returns></returns>
+        static XamlWindow& Make(PCWSTR id, winrt::Windows::UI::Xaml::UIElement(*getUI)(const XamlWindow&), AppController* c = nullptr) {
+            XamlWindow xw(id);
+            xw.controller = c;
+            xw.getUI = getUI;
+            const auto& entry = s_windows.emplace(id, std::move(xw));
+            return entry.first->second;
+        }
+
+        std::wstring_view Id() const {
+            return m_Id;
+        }
+        static XamlWindow& Get(PCWSTR id) {
+            return s_windows.at(id);
+        }
+        static constexpr PCWSTR WindowClass() {
+            return L"XamlWindow";
+        }
+
+        HWND Create(PCWSTR szTitle, DWORD style, HWND parent = nullptr, int width = CW_USEDEFAULT, int height = CW_USEDEFAULT, int nCmdShow = SW_NORMAL) {
+            static std::once_flag once;
+
+            std::call_once(once, [ctl = this->controller]() {
+                WNDCLASSEXW wcex{};
+                wcex.cbSize = sizeof(wcex);
+
+                wcex.style = CS_HREDRAW | CS_VREDRAW;
+                wcex.lpfnWndProc = [](HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam) -> LRESULT
+                {
+                    // Get handle to the core window.
+                    auto xw = reinterpret_cast<XamlWindow*>(GetWindowLongPtr(hWnd, GWLP_USERDATA));
+                    if (xw == nullptr && message == WM_CREATE) {
+                        auto createStruct = reinterpret_cast<LPCREATESTRUCT>(lParam);
+
+                        xw = reinterpret_cast<XamlWindow*>(createStruct->lpCreateParams);
+                        SetWindowLongPtr(hWnd, GWLP_USERDATA, reinterpret_cast<LONG_PTR>(xw));
+                    }

```

Comment:
not strictly needed, but consider clearing the pointer in WM_NCDESTROY (the last window message)